### PR TITLE
storage: configure baremetal to use EmptyDir

### PR DESCRIPTION
**Pull request by honza**

### CODE REVIEW

**Comment by bparees** [comment]
bootstrapping as removed, not emptydir, was a deliberate choice so users aren't surprised when they lose all their stored data because they were unknowingly running on emptydir storage.  What's the justification for this change?

**Comment by honza** [comment]
Sorry for the bare PR! A bunch of our tests are broken because the image registry is turned off.

**Comment by bparees** [comment]
> Sorry for the bare PR! A bunch of our tests are broken because the image registry is turned off.

what tests?  what scenario?

In general i'd expect the solution here to be that you need to do some manual steps in your test job prior to running your tests(enable the registry), not change how the product works.

**Comment by stbenjam** [comment]
@bparees  I know this was intentional way back in the early days of baremetal IPI. However, this PR is coming after a discussion with @smarterclayton. We could enable it just for the tests, but the larger problem is the baremetal platform as a GA platform needs to have the core features of OCP working.

Could we enable EmptyDir and provide some documentation about the pitfalls of using it on baremetal? Production users would probably want to reconfigure it to use something like OCS or local storage operator, but having it work out of the box is important.

File: pkg/storage/storage.go [hunk 1]
```diff
@@ -167,6 +165,8 @@ func GetPlatformStorage(listers *regopclient.Listers) (imageregistryv1.ImageRegi
 	// EmptyDir storage.
 	case configapiv1.LibvirtPlatformType:
 		fallthrough
+	case configapiv1.BareMetalPlatformType:
+		fallthrough
```
Comment (stbenjam) [inline pkg/storage/storage.go:169]: The comment on L166 needs an update.  Or possible add a comment here explaining the justification for why we use EmptyDir on initial baremetal IPI install.

**Comment by bparees** [comment]
> Could we enable EmptyDir and provide some documentation about the pitfalls of using it on baremetal? Production users would probably want to reconfigure it to use something like OCS or local storage operator, but having it work out of the box is important.

i'd like to know what has changed since this last time we had these discussions.  Documentation is not sufficient imho.  There are two major negative implications that result from defaulting clusters in this way:

1) they will lose their image content if the pod is rescheduled
2) they can't easily migrate to a persistent storage, again without losing their existing content

What is the benefit of giving them a "working" registry that they can't actually use for anything because it's not persistent(and also not HA)?

**Comment by smarterclayton** [comment]
We should be testing this way (emptydir) on baremetal via CI provided config before we change the core product.

**Comment by bparees** [comment]
> We could enable it just for the tests, but the larger problem is the baremetal platform as a GA platform needs to have the core features of OCP working.

to state it somewhat more succinctly, i don't consider a registry running on emptydir storage to be "a core feature of OCP" (it's insufficient)

I pinged Clayton about this and the net is:

1) for now you should fix your test job to setup the registry/storage for your tests
2) if we did anything like this, we should use localvolume, not emptydir (provides at least some more persistence guarantees)
3) even better we should consider a "no storage"/read-only registry option (rejects pushes, doesn't cache pullthrough content).

**Comment by stbenjam** [comment]
> i'd like to know what has changed since this last time we had these discussions. 

A lot... we're GA, for one.  When we had these discussions we were planning on always using Rook and this was going to be just an install time thing that gets reconfigured shortly after installation.

> to state it somewhat more succinctly, i don't consider a registry running on emptydir storage to be "a core feature of OCP" (it's insufficient)

Not having a registry at all is worse. If a customer deploys a baremetal PoC and then has to do a bunch of things after that to make it work like other IPI clusters on a cloud platform, that sucks as an experience.

On-premise production deployments of OCP are different, and there's always going to need to be some "production deployment" considerations for baremetal that customers need to understand.

> if we did anything like this, we should use localvolume, not emptydir (provides at least some more persistence guarantees)

Could we do this instead? Or are you totally against turning on the registry at all for baremetal by default?

**Comment by bparees** [comment]
> Not having a registry at all is worse.

I consider a customer being surprised to lose all their data as far worse than not having a registry(or having to take some manual steps to enable one).

> Could we do this instead? Or are you totally against turning on the registry at all for baremetal by default?

Personally i'm still not a big fan, but i'd probably be willing to compromise on it.  Previously when we went down this route, we said we'd provide an alerting event to the admin that their registry storage was not persistent. Ultimately there was pushback on that alert(that it would be annoying to admins who knew and didn't care, as well as the concern that we'd be configuring them out of the box with an alert saying the cluster wasn't configured properly) which lead to us saying we're just not configure this way at all except for non-production scenarios(meaning CodeReadyContainers).

If we're revisiting enabling the registry in a less than production ready fashion, I think we also need to revisit how we ensure admins are aware that they are in such a mode.  I don't think just putting something in the docs is sufficient.

> When we had these discussions we were planning on always using Rook and this was going to be just an install time thing that gets reconfigured shortly after installation.

so what happened to this?

**Comment by stbenjam** [comment]
> > Not having a registry at all is worse.
> 
> I consider a customer being surprised to lose all their data as far worse than not having a registry(or having to take some manual steps to enable one).
> 
> > Could we do this instead? Or are you totally against turning on the registry at all for baremetal by default?
> 
> Personally i'm still not a big fan, but i'd probably be willing to compromise on it. Previously when we went down this route, we said we'd provide an alerting event to the admin that their registry storage was not persistent. Ultimately there was pushback on that alert(that it would be annoying to admins who knew and didn't care, as well as the concern that we'd be configuring them out of the box with an alert saying the cluster wasn't configured properly) which lead to us saying we're just not configure this way at all except for non-production scenarios(meaning CodeReadyContainers).
> 
> If we're revisiting enabling the registry in a less than production ready fashion, I think we also need to revisit how we ensure admins are aware that they are in such a mode. I don't think just putting something in the docs is sufficient.

Ok -  I think you've convinced me to not do this for now. We'll turn on EmptyDir just in CI, that'll let us solve the immediate problem of getting more tests to pass. 

> > When we had these discussions we were planning on always using Rook and this was going to be just an install time thing that gets reconfigured shortly after installation.
> 
> so what happened to this?

We became less opinionated, I guess.  The original idea was far more prescriptive than the current platform has become. This is probably better suited to a conversation with product people than with me on GitHub, though.

**Comment by honza** [comment]
Thanks!

**Pull request closed without merging**